sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -86,13 +86,24 @@ import org.apache.spark.sql.types._
  * Further, the above `if(isnull(fromExp), null, false)` is represented using conjunction
  * `and(isnull(fromExp), null)`, to enable further optimization and filter pushdown to data sources.
  * Similarly, `if(isnull(fromExp), null, true)` is represented with `or(isnotnull(fromExp), null)`.
+ *
+ * Unwrap casts in In expression with patterns like:
```

Comment:
I'm wondering if we can add this to the beginning of the doc:

```
 * Unwrap casts in binary comparison or `In` operations with patterns like following:
 *
 * - `BinaryComparison(Cast(fromExp, toType), Literal(value, toType))`
 * - `BinaryComparison(Literal(value, toType), Cast(fromExp, toType))`
 * - `In(Cast(fromExp, toType), Seq((v1, toType), (v2, toType), ...)`
```

Then a separate section with details how the optimization is done for `In` clauses.